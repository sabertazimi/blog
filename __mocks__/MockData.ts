const time = new Date(2022, 0, 1, 8, 0, 0).toLocaleString('zh-CN', {
  hour12: false,
});

const siteMetadata = {
  title: 'Title',
  author: 'Sabertaz',
  siteUrl: 'https://example.com',
  email: 'example@github.com',
  disqusUrl: 'https://example.com',
  landingTitles: ['A', 'B', 'C'],
  socialList: {
    github: 'sabertazimi',
    twitter: 'sabertazimi',
    facebook: 'sabertazimi',
    linkedin: 'sabertazimi',
    weibo: 'sabertazimi',
  },
  bookList: Array.from(Array(3).keys()).map(() => ({
    title: 'Title',
    author: 'Sabertaz',
    url: 'https://example.com',
    description: 'Description',
  })),
};

const basePosts = Array.from(Array(5).keys()).map(index => ({
  index,
  fields: { slug: `/${index + 1}BasicNotes/` },
  frontmatter: { title: `${index + 1} Basic Notes` },
  timeToRead: index + 1,
}));

const posts = basePosts.map((post, index) => ({
  ...post,
  fields: { slug: `/${index + 1}BasicNotes/`, gitTime: `${index + 1} time` },
  frontmatter: {
    title: `${index + 1} Basic Notes`,
    subtitle: 'Be a Stupid Learner',
    author: 'Sabertaz',
    date: '2018-08-08T00:00:00.000Z',
    tags: ['JavaScript', 'Frontend Development', 'Web Development'],
  },
  timeToRead: index + 1,
}));

const MockData = {
  time,
  siteMetadata,
  basePosts,
  posts,
};

export default MockData;
