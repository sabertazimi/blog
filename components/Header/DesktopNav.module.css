.primary[class] {
  @apply bg-gradient-primary shadow-none;
}

.transparent[class] {
  @apply bg-white/30 shadow-lg;
}

.light.light[class],
.dark.dark[class] {
  @apply hover:text-primary focus:text-primary;
}

.light.light[class] {
  @apply text-light;
}

.dark.dark[class] {
  @apply text-dark;
}
