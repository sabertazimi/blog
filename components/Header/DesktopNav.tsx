import Link from '@components/Link'
import LocalImage from '@components/LocalImage'
import Menu from '@components/Menu'
import PostsSearchBar from '@components/PostsSearchBar'
import ThemeSwitch from '@components/ThemeSwitch'
import { cx } from '@components/utils'
import { routes } from '@config'
import logo from '@images/logo-full.png'
import type { PostMeta } from '@types'
import styles from './DesktopNav.module.css'

interface Props {
  fixed: boolean
  posts: PostMeta[]
}

function DesktopNav({ fixed, posts }: Props): JSX.Element {
  return (
    <Menu
      mode="horizontal"
      className={cx(
        'fixed top-0 z-10 hidden w-full transform-gpu',
        'border-transparent bg-transparent font-extrabold backdrop-blur-md transition',
        'md:visible md:flex md:justify-center',
        {
          [styles.primary]: !fixed,
          [styles.transparent]: fixed,
        },
      )}
      items={[
        {
          key: 'home',
          label: (
            <Link href="/" className="flex-container transform-gpu transition">
              <LocalImage src={logo} alt="Logo" width={96} height={96} />
            </Link>
          ),
        },
        ...routes.map(route => ({
          key: route.id,
          label: (
            <Link
              href={route.path}
              className={cx(
                'flex-container dark-menu-link h-full transform-gpu transition',
                {
                  [styles.light]: !fixed,
                  [styles.dark]: fixed,
                },
              )}
            >
              {route.name}
            </Link>
          ),
        })),
        {
          key: 'search',
          label: (
            <span className="flex-container h-full transform-gpu border-transparent transition md:w-52 lg:w-60 xl:w-max">
              <PostsSearchBar posts={posts} />
            </span>
          ),
        },
        {
          key: 'theme',
          label: (
            <ThemeSwitch
              className={cx('dark-menu-link', {
                [styles.light]: !fixed,
                [styles.dark]: fixed,
              })}
            />
          ),
        },
      ]}
    />
  )
}

export default DesktopNav
