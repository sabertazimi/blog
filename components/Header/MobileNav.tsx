import Col from '@/components/Col'
import { Bars } from '@/components/Icons'
import Link from '@/components/Link'
import LocalImage from '@/components/LocalImage'
import Menu from '@/components/Menu'
import Popover from '@/components/Popover'
import Row from '@/components/Row'
import ThemeSwitch from '@/components/ThemeSwitch'
import { routes } from '@/config'
import logo from '@/images/logo-full.png'

function MobileNav(): JSX.Element {
  return (
    <Row className="md:hidden" justify="center" align="middle">
      <Col className="flex-container" span={8}>
        <Popover
          placement="bottomRight"
          trigger="click"
          content={(
            <Menu
              mode="vertical"
              className="w-full min-w-xs max-w-md text-center text-2xl font-extrabold"
              items={[
                ...routes.map(route => ({
                  key: route.id,
                  label: <Link href={route.path}>{route.name}</Link>,
                })),
              ]}
            />
          )}
        >
          <Bars className="text-5xl text-primary" />
        </Popover>
      </Col>
      <Col className="flex-container" span={8}>
        <Link href="/">
          <LocalImage src={logo} alt="Logo" width={96} height={96} />
        </Link>
      </Col>
      <Col className="flex-container" span={8}>
        <ThemeSwitch className="dark:text-light" />
      </Col>
    </Row>
  )
}

export default MobileNav
