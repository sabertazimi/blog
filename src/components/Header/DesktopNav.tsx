import { PostsSearchBar } from '@components';
import { Routes } from '@config';
import type { PostMetaType } from '@types';
import { Menu } from 'antd';
import classNames from 'classnames';
import { Link } from 'gatsby';
import logo from 'images/logo-full.png';
import React from 'react';

interface Props {
  fixed: boolean;
  posts: PostMetaType[];
}

const DesktopNav = ({ fixed, posts }: Props): JSX.Element => (
  <Menu
    mode="horizontal"
    className={classNames(
      'fixed z-100 w-full font-extrabold bg-transparent border-transparent transition transform-gpu hidden md:visible md:flex',
      {
        'bg-gradient-primary': !fixed,
        'bg-white/30': fixed,
        'shadow-lg': fixed,
        'backdrop-blur-primary': fixed,
      }
    )}
  >
    <Menu.Item
      key="home"
      className={classNames('flex-container transition transform-gpu', {
        'scale-100': !fixed,
        'scale-0': fixed,
      })}
    >
      <Link to="/">
        <img className="block w-full h-24" src={logo as string} alt="Logo" />
      </Link>
    </Menu.Item>
    {Routes.map(route => (
      <Menu.Item
        key={route.id}
        className={classNames('flex-container transition transform-gpu', {
          'translate-x-0': !fixed,
          '-translate-x-28': fixed,
        })}
      >
        <Link
          className={classNames('transition-none', {
            'text-light': !fixed,
            'text-dark': fixed,
          })}
          to={route.path}
        >
          {route.name}
        </Link>
      </Menu.Item>
    ))}
    <Menu.Item
      key="search"
      className={classNames(
        'flex-container border-transparent transition transform-gpu',
        {
          'translate-x-0': !fixed,
          '-translate-x-28': fixed,
        }
      )}
    >
      <PostsSearchBar posts={posts} />
    </Menu.Item>
  </Menu>
);

export default DesktopNav;
