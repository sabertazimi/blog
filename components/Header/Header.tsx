import { useVisibility } from '@hooks'
import type { PostMeta } from '@types'
import { useCallback, useRef, useState } from 'react'
import DesktopNav from './DesktopNav'
import MobileNav from './MobileNav'

interface Props {
  posts: PostMeta[]
}

const Header = ({ posts }: Props): JSX.Element => {
  const [navFixed, setNavFixed] = useState(false)
  const headerRef = useRef<HTMLDivElement>(null)

  const hideFixedNav = useCallback(() => {
    setNavFixed(false)
  }, [])

  const showFixedNav = useCallback(() => {
    setNavFixed(true)
  }, [])

  useVisibility({
    ref: headerRef,
    onBottomPassed: showFixedNav,
    onBottomPassedReverse: hideFixedNav,
  })

  return (
    <div role="grid" ref={headerRef}>
      <MobileNav />
      <DesktopNav fixed={navFixed} posts={posts} />
    </div>
  )
}

export default Header
