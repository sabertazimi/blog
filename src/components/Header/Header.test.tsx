import MockData from '@MockData';
import React from 'react';
import { create } from 'react-test-renderer';
import Header from './Header';

describe('Header', () => {
  const mockPosts = MockData.posts;

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test('should render correctly (snapshot)', () => {
    const tree = create(<Header posts={mockPosts} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
