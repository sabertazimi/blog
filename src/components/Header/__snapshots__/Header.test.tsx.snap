// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly (snapshot) 1`] = `
<div>
  <div
    className="ant-row ant-row-center ant-row-middle md:hidden"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-8 ant-col-offset-8 flex-container"
      style={Object {}}
    >
      <a
        href="/"
      >
        <img
          alt="Logo"
          className="block w-24 h-24"
          src="test-file-stub"
        />
      </a>
    </div>
    <div
      className="ant-col ant-col-8 flex-container"
      style={Object {}}
    >
      <span
        aria-label="bars"
        className="anticon anticon-bars text-5xl text-primary"
        onClick={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="img"
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          data-icon="bars"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M912 192H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM104 228a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0z"
          />
        </svg>
      </span>
    </div>
  </div>
  <ul
    className="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-light fixed z-100 w-full font-extrabold transition transform-gpu border-transparent hidden md:visible md:flex bg-gradient-primary"
    data-menu-list={true}
    onKeyDown={[Function]}
    role="menu"
    tabIndex={0}
  >
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container transition transform-gpu scale-100"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 0,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <a
          href="/"
        >
          <img
            alt="Logo"
            className="block w-full h-24"
            src="test-file-stub"
          />
        </a>
      </span>
    </li>
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container transition transform-gpu translate-x-0"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 1,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <a
          className="transition-none text-light"
          href="/posts"
        >
          Posts
        </a>
      </span>
    </li>
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container transition transform-gpu translate-x-0"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 2,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <a
          className="transition-none text-light"
          href="/tags"
        >
          Tags
        </a>
      </span>
    </li>
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container transition transform-gpu translate-x-0"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 3,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <a
          className="transition-none text-light"
          href="/books"
        >
          Books
        </a>
      </span>
    </li>
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container transition transform-gpu translate-x-0"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 4,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <a
          className="transition-none text-light"
          href="/about"
        >
          About
        </a>
      </span>
    </li>
    <li
      className="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child flex-container border-transparent transition transform-gpu translate-x-0"
      data-menu-id={null}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "height": undefined,
          "opacity": 1,
          "order": 5,
          "overflowY": undefined,
          "pointerEvents": undefined,
          "position": undefined,
        }
      }
      tabIndex={-1}
    >
      <span
        className="ant-menu-title-content"
      >
        <div
          className="ant-select ant-select-auto-complete flex-container ant-select-single ant-select-customize-input ant-select-show-search"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
        >
          <div
            className="ant-select-selector"
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <span
              className="ant-select-selection-search"
            >
              <span
                className="ant-input-affix-wrapper ant-select-selection-search-input"
                onMouseUp={[Function]}
                style={
                  Object {
                    "opacity": null,
                  }
                }
              >
                <span
                  className="ant-input-prefix"
                >
                  <span
                    aria-label="search"
                    className="anticon anticon-search"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="search"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                      />
                    </svg>
                  </span>
                </span>
                <input
                  aria-activedescendant="undefined_list_0"
                  aria-autocomplete="list"
                  aria-controls="undefined_list"
                  aria-haspopup="listbox"
                  aria-owns="undefined_list"
                  autoComplete="off"
                  className="ant-input"
                  hidden={null}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onCompositionEnd={[Function]}
                  onCompositionStart={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onPaste={[Function]}
                  placeholder="Search Posts ..."
                  readOnly={false}
                  role="combobox"
                  style={null}
                  type="search"
                  unselectable={null}
                  value=""
                />
                <span
                  className="ant-input-suffix"
                >
                  <span
                    className="ant-input-clear-icon ant-input-clear-icon-hidden"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    role="button"
                    tabIndex={-1}
                  >
                    <span
                      aria-label="close-circle"
                      className="anticon anticon-close-circle"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="close-circle"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </span>
            </span>
            <span
              className="ant-select-selection-placeholder"
            />
          </div>
        </div>
      </span>
    </li>
    <li
      aria-hidden={true}
      className="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="none"
      style={
        Object {
          "height": 0,
          "opacity": 0,
          "order": 9007199254740991,
          "overflowY": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
        }
      }
    >
      <div
        aria-controls={null}
        aria-expanded={false}
        aria-haspopup={true}
        className="ant-menu-submenu-title"
        data-menu-id={null}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={null}
        tabIndex={-1}
        title={null}
      >
        <span
          aria-label="ellipsis"
          className="anticon anticon-ellipsis"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
        <i
          className="ant-menu-submenu-arrow"
        />
      </div>
    </li>
  </ul>
  <div
    aria-hidden={true}
    style={
      Object {
        "display": "none",
      }
    }
  />
</div>
`;
